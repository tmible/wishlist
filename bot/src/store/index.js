import migrate from '@tmible/wishlist-common/db-migrations';
import { provide } from '@tmible/wishlist-common/dependency-injector';
import Database from 'better-sqlite3';
import InjectionToken from '@tmible/wishlist-bot/architecture/injection-token';
import EditingModule from './editing/index.js';
import UsernamesModule from './usernames/index.js';
import WishlistModule from './wishlist/index.js';

/**
 * Обязательная часть элемента разметки текста
 * @typedef {object} EntityBase
 * @property {string} type Тип элемента
 * @property {number} offset Отступ от начала строки
 * @property {number} length Длина
 */
/**
 * Элемент разметки текста
 * @typedef {EntityBase & Record<string, unknown>} Entity
 */

/** @module Хранилище -- абстракция БД */

/**
 * Модули хранилища
 * @type {{ configure: Function } & Record<string, unknown>}
 */
const storeModules = [
  WishlistModule,
  EditingModule,
  UsernamesModule,
];

/**
 * Создание подключения к БД, [миграции БД]{@link migrate} и настройка модулей хранилища
 * @function initStore
 * @returns {() => void} Функция закрытия подключения к БД
 */
const initStore = () => {
  const db = new Database(process.env.WISHLIST_DB_FILE_PATH);
  migrate(db, process.env.WISHLIST_DB_MIGRATIONS_PATH);
  provide(InjectionToken.Database, db);
  storeModules.forEach(({ configure }) => configure());
  return () => db.close();
};

export default initStore;
