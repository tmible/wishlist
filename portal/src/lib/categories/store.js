import { initCategories } from '$lib/categories/use-cases/init-categories.js';

/** @typedef {import('./domain.js').Category} Category */

/** @module Адаптер хранилища категорий */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {Category[]}
 */
let value = [];

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * категорий
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => Category[]} get Получение значения хранилища
 * @property {(newValue: Category[]) => void} set Установка значения хранилища
 * @property {(category: Category) => void} add Добавление категории в хранилище
 * @property {(category: Category) => void} update Обновление категории в хранилище
 * @property {(category: Category) => void} delete Удаление категории из хранилища
 */
export const categories = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);

    if (subscriptions.size === 1) {
      initCategories();
    }

    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  add: (category) => {
    value.push(category);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  update: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value[categoryIndex] = category;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  delete: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value.splice(categoryIndex, 1);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
