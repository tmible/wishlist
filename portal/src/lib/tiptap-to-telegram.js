/** @module Перевод дерева текстового редактора в набор элементов разметки текста телеграма */

/** @typedef {import('@tiptap/core').JSONContent} JSONContent */
/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */
/**
 * @template ArrayType
 * @typedef
 * {ArrayType extends readonly (infer ElementType)[] ? ElementType : never}
 * ArrayElement<ArrayType extends readonly unknown[]>
 */
// eslint-disable-next-line no-secrets/no-secrets -- Просто длинное название
/**
 * Уплощёный узел дерева текстового редактора
 * @typedef {object} FlatChild
 * @property {string[]} marks Набор активных меток
 * //Будет работать только в TS
 * //@property {Pick<ArrayElement<Pick<JSONContent, 'marks'>>, 'type'>[]} marks Набор активных меток
 * @property {Pick<JSONContent, 'text'>} text Содержащийся в узле текст
 * @property {Pick<JSONContent, 'attrs'>} attrs Атрибуты для меток
 */

/**
 * Отображение названий меток текстового редактора в типы элементов
 * разметки текста телегарма Не упомянутые совпадают
 * @constant {Map<string, Pick<Entity, 'type'>>}
 */
const TIPTAP_TO_TELEGRAM_NAMES_MAP = new Map([
  [ 'strike', 'strikethrough' ],
  [ 'codeBlock', 'pre' ],
  [ 'link', 'url' ],
]);

/**
 * Рекурсивное уплощение дерева текстового редактора
 * @function flatContent
 * @param {JSONContent} content Текущий узел дерева
 * @param {string[]} [mixin] Примесь меток от родительских узлов дерева
 * @returns {FlatChild[]} Массив уплощённых потомков текущего узла
 */
const flatContent = (content, mixin = []) => content?.flatMap((child) => {
  if (child.type === 'text') {
    const url = child.marks?.find(({ type }) => type === 'text_link')?.attrs.href;
    return {
      marks: [ ...mixin, ...(child.marks?.map(({ type }) => type) ?? []) ],
      text: child.text,
      ...(url ? { attrs: { url } } : {}),
    };
  }

  if (child.type === 'codeBlock') {
    return {
      marks: [ child.type ],
      text: child.content[0].text ?? '',
      attrs: child.attrs,
    };
  }

  if (child.type === 'blockquote') {
    return flatContent(child.content, [ ...mixin, child.type ]);
  }

  const childContentFlat = flatContent(child.content, mixin) ?? [];
  if (child.content) {
    childContentFlat.at(-1).text += '\n';
  }
  return childContentFlat;
});

/**
 * Дополнительные проверки для приведения к соответствию формату Телеграма
 * @function preAndTextLinksMapper
 * @param {Entity} entity Элемент разметки текста
 * @returns {Entity} Обновлённый элемент разметки текста
 */
const preAndTextLinksMapper = (entity) => {
  if (entity.type === 'pre' && entity.language === null) {
    delete entity.language;
  }
  if (entity.type === 'text_link' && !entity.url.endsWith('/')) {
    entity.url += '/';
  }
  return entity;
};

/**
 * Перевод дерева текстового редактора в набор элементов разметки текста телеграма
 * 1. [Уплощение дерева]{@link flatContent};
 * 2. Перебор получившихся блоков:
 * 2.1. Создание элементов разметки текста на основе меток, закончившихся на начале текущего блока;
 * 2.2. Добавление к активным меткам меток, начавшихся в начале текущего блока;
 * 2.3. Добавление к тексту текста текущего блока;
 * 3. Удаление пробельных символов с конца текста;
 * 4. Создание элементов разметки текста на основе меток, закончившихся в конце текста;
 * @function tiptapToTelegram
 * @param {JSONContent} content Корень дерева текстового редактора
 * @returns {[ string, Entity[] ]} Текст и набор элементов его разметки
 */
export const tiptapToTelegram = (content) => {
  const contentFlat = flatContent(content);

  let text = '';
  const entities = [];
  let activeMarks = [];

  for (
    let i = 0, finishedMarks = [], [ block ] = contentFlat;
    i < contentFlat.length;
    ++i, text += block.text, block = contentFlat[i]
  ) {
    ({ activeMarks = [], finishedMarks = [] } = Object.groupBy(
      activeMarks,
      ({ type }) => (block.marks.includes(type) ? 'activeMarks' : 'finishedMarks'),
    ));

    entities.push(
      ...finishedMarks
        .map((finishedMark) => ({
          ...finishedMark,
          type: TIPTAP_TO_TELEGRAM_NAMES_MAP.get(finishedMark.type) ?? finishedMark.type,
          length: text.length - finishedMark.offset,
        }))
        .map((entity) => preAndTextLinksMapper(entity)),
    );

    activeMarks.push(
      ...block.marks
        .filter((mark) => !activeMarks.some(({ type }) => type === mark))
        .map((mark) => ({ ...block.attrs, offset: text.length, type: mark })),
    );
  }

  text = text.trimEnd();

  entities.push(
    ...activeMarks
      .map((activeMark) => ({
        ...activeMark,
        type: TIPTAP_TO_TELEGRAM_NAMES_MAP.get(activeMark.type) ?? activeMark.type,
        length: text.length - activeMark.offset,
      }))
      .map((entity) => preAndTextLinksMapper(entity)),
  );

  return [ text, entities ];
};
