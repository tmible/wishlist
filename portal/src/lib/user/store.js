import { DEFAULT_USER } from './default-user.const.js';

/** @typedef {import('./domain.js').User} User */

/** @module Адаптер хранилища пользователя */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {User}
 */
let value = DEFAULT_USER;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * пользователя
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => User} get Получение значения хранилища
 * @property {(newValue: User) => void} set Установка значения хранилища
 */
export const user = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },
  get: () => value,
  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
