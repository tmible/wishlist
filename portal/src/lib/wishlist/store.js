/** @typedef {import('./domain.js').OwnWishlistItem} OwnWishlistItem */
/** @typedef {import('./use-cases/reorder-list.js').ReorderPatch} ReorderPatch */

/** @module Адаптер хранилища списка желаний */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {OwnWishlistItem[] | null}
 */
let value = null;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * Уведомление подписчиков об изменении значения хранилища
 * @function notifySubscribers
 * @returns {void}
 */
const notifySubscribers = () => {
  for (const subscriber of subscriptions.values()) {
    subscriber(value);
  }
};

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * списка желаний
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => OwnWishlistItem[]} get Получение значения хранилища
 * @property {(newValue: OwnWishlistItem[]) => void} set Установка значения хранилища
 * @property {(item: OwnWishlistItem) => void} add Добавление элемента списка в хранилище
 * @property {(item: OwnWishlistItem) => void} update Обновление элемента списка в хранилище
 * @property {(itemIds: OwnWishlistItem['id'][]) => void} delete
 *   Удаление элементов списка из хранилища
 * @property {(patch: ReorderPatch) => void} reorder Переупорядочивание списка в хранилище
 */
export const wishlist = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    notifySubscribers();
  },

  add: (item) => {
    if (!value) {
      value = [];
    }
    value.push(item);
    notifySubscribers();
  },

  update: (item) => {
    const itemIndex = value?.findIndex(({ id }) => id === item.id);
    if (itemIndex === -1) {
      return;
    }
    value[itemIndex] = item;
    notifySubscribers();
  },

  delete: (itemIds) => {
    let isAnyDeleted = false;

    for (const id of itemIds) {
      const itemIndex = value?.findIndex((item) => id === item.id);
      if (itemIndex === -1) {
        continue;
      }
      value.splice(itemIndex, 1);
      isAnyDeleted = true;
    }

    if (!isAnyDeleted) {
      return;
    }

    notifySubscribers();
  },

  reorder: (patch) => {
    if (!value) {
      return;
    }
    for (const item of value) {
      const reorder = patch.find(({ id }) => id === item.id);
      item.order = reorder?.order ?? item.order;
    }
    value.sort((a, b) => a.order - b.order);
    notifySubscribers();
  },
};
