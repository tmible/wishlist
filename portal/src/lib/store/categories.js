import { writable } from 'svelte/store';

/**
 * @template T
 * @typedef {import('svelte/store').Writable<T>} Writable
 */
/**
 * Категория элементов списка желаний пользователя
 * @typedef {object} Category
 * @property {number} id Идентификатор категории
 * @property {string} name Название категории
 */

/**
 * Svelte хранилище категорий элементов списка желаний пользователя
 * @type {Writable<Category[] | null>}
 */
export const categories = writable(null, (set, update) => {
  let timeout;

  update((value) => {
    if (value === null) {
      timeout = setTimeout(async () => set(await fetch(
        '/api/wishlist/categories',
      ).then(
        (response) => response.json(),
      )));
    }
    return value;
  });

  return () => timeout && clearTimeout(timeout);
});
