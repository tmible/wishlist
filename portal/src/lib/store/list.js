import { writable } from 'svelte/store';

/**
 * @template T
 * @typedef {import('svelte/store').Writable<T>} Writable
 */
/** @typedef {import('@tmible/wishlist-common/constants/list-item-state').default} ListItemState */
/** @typedef {import('$lib/components/telegram-entities/parser.svelte').Entity} Entity */
/** @typedef {import('./categories.js').Category} Category */
/**
 * Элемент списка желаний пользователя во варианте отображения для владельца
 * @typedef {object} OwnListItem
 * @property {number} id Идентификатор элемента
 * @property {string} name Название подарка
 * @property {string} description Описание подарка
 * @property {ListItemState} state Состояние подарка
 * @property {number} order Порядковый номер элемента
 * @property {Category | null} category Категория элемента
 * @property {Entity[]} descriptionEntities Элементы разметки текста описания подарка
 */

/**
 * Svelte хранилище списка желаний пользователя
 * @type {Writable<OwnListItem[] | null>}
 */
export const list = writable(null);
