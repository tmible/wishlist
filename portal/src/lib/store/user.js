import { writable } from 'svelte/store';

/**
 * @template T
 * @typedef {import('svelte/store').Writable<T>} Writable
 */

/**
 * Svelte хранилище идентификатора и состояния аутентифицированности пользователя
 * @type {Writable<{ id: string | null; isAuthenticated: boolean | null }>}
 */
export const user = writable({ id: null, isAuthenticated: null });
