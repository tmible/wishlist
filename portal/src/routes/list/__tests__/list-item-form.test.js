// @vitest-environment jsdom
import { cleanup, fireEvent, render, screen, waitFor } from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import arrayToOrderedJSON from '@tmible/wishlist-common/array-to-ordered-json';
import { writable } from 'svelte/store';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { tiptapToTelegram } from '$lib/tiptap-to-telegram.js';

const dispatch = vi.fn();

vi.stubGlobal('fetch', vi.fn(() => ({ ok: true })));
vi.mock('@tmible/wishlist-common/array-to-ordered-json');
vi.mock('$lib/tiptap-to-telegram.js');
vi.mock(
  '$lib/store/list',
  () => ({
    list: writable([{
      id: 'id',
      description: 'description',
      descriptionEntities: [{ offset: 0, type: 0 }],
      property1: 'stored1',
      property2: 'stored2',
    }]),
  }),
);
vi.mock('$lib/store/user', () => ({ user: writable({ id: 'userid' }) }));

describe('list item form', () => {
  let ListItemForm;

  beforeEach(async () => {
    vi.doMock('svelte');
    vi.mocked(
      await import('svelte').then(({ createEventDispatcher }) => createEventDispatcher),
    ).mockReturnValue(dispatch);
    vi.doMock(
      '$lib/components/text-editor',
      async () => ({ default: await import('./mock.svelte').then((module) => module.default) }),
    );
    ListItemForm = await import('../list-item-form.svelte').then((module) => module.default);
  });

  afterEach(() => {
    vi.resetModules();
    vi.clearAllMocks();
    cleanup();
  });

  it('should be displayed with empty inputs', () => {
    const { container } = render(ListItemForm);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should prefill inputs', () => {
    render(
      ListItemForm,
      {
        values: {
          name: 'name',
          description: 'description',
          priority: 10,
        },
      },
    );
    expect([
      ...screen.getAllByRole('textbox'),
      ...screen.getAllByRole('spinbutton'),
    ].map(({ value }) => value)).toMatchSnapshot();
  });

  it('should dispatch cancel event on cancel button click', async () => {
    const user = userEvent.setup();
    render(ListItemForm);
    await user.click(screen.getByText('Отмена', { selector: 'button' }));
    expect(dispatch).toHaveBeenCalledWith('cancel');
  });

  describe('on submit button click', () => {
    let component;
    let form;
    let formData;

    beforeEach(() => {
      formData = {
        get: vi.fn(() => JSON.stringify({ content: 'content' })),
        set: vi.fn(),
        append: vi.fn(),
        delete: vi.fn(),
        entries: vi.fn(() => []),
      };
      vi.spyOn(Object, 'fromEntries').mockReturnValue({});
      vi.stubGlobal('FormData', vi.fn(() => formData));
      vi.mocked(tiptapToTelegram).mockReturnValue([]);
      let baseElement;
      ({ baseElement, component } = render(ListItemForm));
      form = baseElement.querySelector('form');
    });

    it('should call tiptapToTelegram', () => {
      fireEvent.submit(form);
      expect(tiptapToTelegram).toHaveBeenCalledWith('content');
    });

    it('should set description', () => {
      tiptapToTelegram.mockReturnValueOnce([ 'description', [ 'entity 1', 'entity 2' ]]);
      fireEvent.submit(form);
      expect(formData.set).toHaveBeenCalledWith('description', 'description');
    });

    it('should set description entities', () => {
      tiptapToTelegram.mockReturnValueOnce([ 'description', [ 'entity 1', 'entity 2' ]]);
      fireEvent.submit(form);
      expect(
        formData.set,
      ).toHaveBeenCalledWith(
        'descriptionEntities',
        JSON.stringify([ 'entity 1', 'entity 2' ]),
      );
    });

    describe('if was prefilled', () => {
      beforeEach(() => {
        component.$set({ values: { id: 'id' } });
        formData.entries.mockReturnValueOnce([
          [ 'property1', 'stored1' ],
          [ 'property2', 'fromForm2' ],
          [ 'descriptionEntities', '[{"type":0,"offset":0}]' ],
        ]);
        vi.mocked(arrayToOrderedJSON).mockReturnValue('[]');
      });

      it('should filter values', () => {
        fireEvent.submit(form);
        expect(formData.delete).toHaveBeenCalledWith('property1');
      });

      it('should filter description entities', () => {
        fireEvent.submit(form);
        expect(formData.delete).toHaveBeenCalledWith('descriptionEntities');
      });

      it('should dispatch cancel event if no values left after filter', () => {
        tiptapToTelegram.mockReturnValueOnce([ 'description', []]);
        fireEvent.submit(form);
        expect(dispatch).toHaveBeenCalledWith('cancel');
      });

      it('should send form via "PUT"', () => {
        formData.entries.mockReturnValueOnce([
          [ 'property1', 'stored1' ],
          [ 'property2', 'fromForm2' ],
        ]);
        vi.spyOn(Object, 'fromEntries').mockReturnValue({ from: 'entries' });
        fireEvent.submit(form);
        expect(vi.mocked(fetch)).toHaveBeenCalledWith(
          '/api/wishlist/id',
          {
            method: 'PUT',
            body: JSON.stringify({ from: 'entries' }),
          },
        );
      });
    });

    it('should append userid', () => {
      fireEvent.submit(form);
      expect(formData.append).toHaveBeenCalledWith('userid', 'userid');
    });

    it('should send form via "POST"', () => {
      vi.spyOn(Object, 'fromEntries').mockReturnValueOnce({ from: 'entries' });
      fireEvent.submit(form);
      expect(vi.mocked(fetch)).toHaveBeenCalledWith(
        '/api/wishlist',
        {
          method: 'POST',
          body: JSON.stringify({ from: 'entries' }),
        },
      );
    });

    it('should dispatch success event on form success', async () => {
      fireEvent.submit(form);
      await waitFor(() => expect(dispatch).toHaveBeenCalledWith('success'));
    });
  });
});
