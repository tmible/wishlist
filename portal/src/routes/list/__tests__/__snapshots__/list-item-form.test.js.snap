// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list item form > should be displayed with empty inputs 1`] = `
<form
  class="prose"
  method="POST"
>
  <label
    class="form-control w-full mb-2"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Название
      </span>
    </div>
     
    <input
      class="input input-bordered w-full bg-base-200"
      name="name"
      placeholder="Название"
      required=""
      type="text"
    />
  </label>
   
  <label
    class="form-control w-full mb-2"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Описание
      </span>
    </div>
     
     
    

Mock
  </label>
   
  <label
    class="form-control w-full mb-6"
  >
    <div
      class="label"
    >
      <span
        class="label-text"
      >
        Категория
      </span>
    </div>
     
    <input
      aria-hidden="true"
      data-melt-select-hidden-input=""
      data-select-input=""
      hidden="true"
      name="categoryId"
      style="position: absolute; opacity: 0; pointer-events: none; margin: 0px; transform: translateX(-100%);"
      tabindex="-1"
    />
    
     
    <button
      aria-autocomplete="list"
      aria-expanded="false"
      class="select select-bordered bg-base-200 items-center"
      data-melt-select-trigger=""
      data-select-trigger=""
      role="combobox"
      type="button"
    >
      <span
        data-select-value=""
      >
        ⸻
      </span>
      
    </button>
    
     
    
  </label>
   
  <div
    class="card-actions"
  >
    <button
      class="btn btn-neutral w-full md:flex-1"
      type="button"
    >
      Отмена
    </button>
     
    <button
      class="btn btn-primary w-full md:flex-1"
      type="submit"
    >
      Добавить
    </button>
  </div>
</form>
`;

exports[`list item form > should prefill inputs 1`] = `
[
  "name",
  "categoryId",
]
`;
